[Feature] 회원가입, 탈퇴를 구현한다.

asd42270 opened the pull request:
# Pull Request 요약

- 회원 가입 탈퇴를 구현한다.

## 변경 사항

### 회원 탈퇴
-  - Member: 논리 삭제를 위해 상속받은 클래스의 delete메서드를 호출하는 삭제 메서드를 추가한다.
- MemberController: 삭제 메서드를 추가한다.
- MemberRegisterRequestDto: MemberRole을 추가한다.
- MemberService: register 메서드에서 member 빌더에 memberRole을 추가한다. withdraw 메서드를 추가한다.
- MemberWithdrawRequestDto: api 명세서에 따라 requsetDto를 추가한다.

### 회원 가입
- MemberRegisterRequestDto: 동, 호수, MemberRole 추가했습니다.
- MemberService: BuildingInfo의 Office에 대한 정보를 어떻게 저장하면 좋을 지에 대해서 논의가 필요할 것 같습니다. 이 부분을 우선 null로 처리하고 회원가입 로직 작성했습니다.
- BuildingInfoRepository: 회원 가입 시 빌딩 정보 저장을 위해 추가했습니다.

## 관련 이슈

- #63 

## 개발 유형

- [ ] 버그 수정
- [x] 새로운 기능 개발
- [ ] 리팩토링
- [x] 테스트 코드 작성
- [ ] 문서 업데이트
- 기타 (아래에 설명 추가)

## 작업 기간

- 작업 시작일: (2024-05-17)
- 작업 종료일: (2024-05-20)

## 유의사항

- 디스코드에서 잠시 얘기했었는데, 자바 코드에서 BuildingInfo의 Office를 저장하는 방식에 대해서 논의가 필요할 것 같습니다!
- 날카로운(?) 코드 리뷰 부탁드립니다ㅋㅋㅋ....달게 받겠습니다!!! 문제가 있는 것 같으나 지식 부족으로 못 찾는 것 같아서...

## 참고문헌

- 이 PR을 작성하며 참조한 문헌이나 가이드가 있다면 여기에 링크를 추가하세요.

asd42270 committed:
[Feat] 회원 탈퇴를 구현한다.
- Member: 논리 삭제를 위해 상속받은 클래스의 delete메서드를 호출하는 삭제 메서드를 추가한다.
- MemberController: 삭제 메서드를 추가한다.
- MemberRegisterRequestDto: MemberRole을 추가한다.
- MemberService: register 메서드에서 member 빌더에 memberRole을 추가한다. withdraw 메서드를 추가한다.
- MemberWithdrawRequestDto: api 명세서에 따라 requsetDto를 추가한다.

asd42270 committed:
[Feat] 회원 탈퇴를 구현한다.
- WithdrawServiceTest: Service 단위 테스트 코드를 작성한다.
- PostMan test 완료

asd42270 committed:
[Feat] 회원 가입을 구현한다.
- MemberRegisterRequestDto: 동, 호수, MemberRole 추가했습니다.
- MemberService: BuildingInfo의 Office에 대한 정보를 어떻게 저장하면 좋을 지에 대해서 논의가 필요할 것 같습니다. 이 부분을 우선 null로 처리하고 회원가입 로직 작성했습니다.
- BuildingInfoRepository: 회원 가입 시 빌딩 정보 저장을 위해 추가했습니다.

asd42270 committed:
[Refactor] 삭제 테스트 코드 수정
- 동 호수 추가

Galmaeki commented on src/main/java/com/final_10aeat/domain/member/controller/MemberController.java:
```diff
@@ -20,17 +19,25 @@ public class MemberController {
+        return ResponseDTO.ok();
+    }
+
+    @DeleteMapping("/delete/v1/users")
```
api 버전을 나누셨는데 앞으로 이렇게 버저닝을 진행하는걸까요?

asd42270 replied:
아 요거 api 명세서에 나온 대로 작업했는데...ㅋㅋㅋㅋㅋ수정하겠습니다!!

Galmaeki commented on src/main/java/com/final_10aeat/domain/member/dto/request/MemberWithdrawRequestDto.java:
```diff
@@ -0,0 +1,7 @@
+package com.final_10aeat.domain.member.dto.request;
+
+public record MemberWithdrawRequestDto(
+        String email,
```
개인적으로는 탈퇴에 이메일은 빼도 될 것 같다는 생각이 듭니다!

ghrltjdtprbs replied:
탈퇴 로직에 현재 로그인한 이메일, 패스워드를 입력하는 과정이 있습니다!

Galmaeki reviewed (changes requested):
고생하셨습니다!
코멘트 남겼으니 확인 부탁드림다!

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/member/controller/MemberController.java:
```diff
@@ -20,17 +19,25 @@ public class MemberController {
+        return ResponseDTO.ok();
+    }
+
+    @DeleteMapping("/delete/v1/users")
```
회원 탈퇴 DELETE /members 
회원가입 POST /members
이렇게만 사용하는게 더 RESTful 할 것 같습니다!
그리고 버저닝도 여러가지 종류가 있기 때문에 버저닝을 도입하려면 더 고민을 해봐야할 것 같습니다! ㅎㅎ

asd42270 replied:
아 넵넵 api 명세서에 나온대로 작업했었는데, 수정하겠습니다!!

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/member/dto/request/MemberRegisterRequestDto.java:
가입시 name도 받아야 합니다!

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/member/service/MemberService.java:
필요없는 주석은 삭제하는게 더 깔끔할 것 같습니다! 그리고 soft delete된 회원 정보들을 필터링하고, 일괄 처리하는 부분도 추가되어야 합니다! 😊

asd42270 replied:
넵!! 주석은 제가 하면서 안 까먹으려고 달아뒀었습니다ㅋㅋㅋ..수정 작업 하면서 다 지우겠습니다!!

ghrltjdtprbs commented on src/test/java/com/final_10aeat/domain/member/unit/WithdrawServiceTest.java:
테스트 케이스가 너무 적은 것 같습니다! 예외 처리한 부분이 잘 동작하는지에 대한 테스트 코드도 추가하면 좋을 것 같아요 👍

asd42270 replied:
넵!! 예외 처리 부분 케이스도 작성하겠습니다잇

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/member/service/MemberService.java:
```diff
@@ -53,6 +70,17 @@ public String login(MemberLoginRequestDto request) {
+                .orElseThrow(MemberNotExistException::new);
+
+        if (!passwordMatcher(request.password(), member)) {
+            throw new MemberNotExistException();
```
MemberNotExistException에서 사용하는 메시지는 "존재하지 않는 사용자입니다." 인데, 로그인한 상태에서 id/pw를 재입력하는 상황에 사용하기에는 적절하지 않은 것 같습니다.
새로운 예외 클래스를 만들거나, MemberNotExistException을 그대로 이용하되, 새로운 메시지를 임의로 커스텀할 수 있도록 MemberNotExistException 코드를 수정하면 좋을 것 같습니다 😊 

ex)
```java
// MemberNotExistException에 추가
public MemberNotExistException(String message) {
        super(ERROR_CODE, message);

// MemberService 수정
throw new MemberNotExistException("현재 로그인한 사용자와 일치하지 않습니다.");
```

ghrltjdtprbs reviewed (changes requested):
수고하셨습니다! 몇가지 리뷰 남겼으니 확인해주세요 👍👍
office 부분은 admin 회원가입이 나오면 이메일 로직 리팩토링을 진행하면서 레디스에 메일을 발송한 admin의 건물 id를 저장할 예정입니다.

asd42270 committed:
[Feat] 회원가입, 탈퇴를 구현한다.
- MemberService: 불필요한 주석 삭제, MemberRepository의 논리삭제 필터링으로 인한 메서드 변경을 반영한다.
- MemberRepository: 논리삭제 필터링을 추가한다.
- MemberRegisterRequestDto: name을 추가한다.
- MemberMissMatchException: 일치하지 않는 사용자에 대한 예외를 추가한다.
- MemberDetailsProvider: MemberRepository의 메서드 변경 사항을 반영한다.
- MemberController: register, withdraw의 api 경로를 수정한다.

asd42270 committed:
[Feat] 회원가입, 탈퇴를 구현한다.
- ErrorCode: 에러 코드를 추가한다.

asd42270 committed:
[Refactor] 테스트코드를 수정한다.
- WithdrawServiceTest:
 - MemberService의 어노테이션을 @InjectMocks로 변경
 - setUp()에서 탈퇴 전 회원가입을 위한 메서드를 작성하고. 이를 위해 BuildingInfoRepository를 Mock으로 등록한다.

Galmaeki commented on src/main/java/com/final_10aeat/domain/member/controller/MemberController.java:
```diff
@@ -1,36 +1,43 @@
     public ResponseDTO<Void> login(
             HttpServletResponse response,
-            @RequestBody MemberLoginRequestDto request) {
+            @RequestBody MemberLoginRequestDto request){
```
서버 최신 사항이 아닌것같슴다!
확인 부탁드려요!

Galmaeki commented on src/main/java/com/final_10aeat/domain/member/controller/MemberController.java:
```diff
@@ -1,36 +1,43 @@
             @RequestBody MemberRegisterRequestDto request
-    ) {
+    ){
         memberService.register(request);
```
정책에서 회원가입에서 로그인까지 바로 됐던걸로 기억합니다!

Galmaeki commented on src/main/java/com/final_10aeat/domain/member/dto/request/MemberRegisterRequestDto.java:
```diff
@@ -1,6 +1,12 @@
 public record MemberRegisterRequestDto(
         String email,
-        String password
+        String password,
```
Validation 달아주심 좋을것같슴다!

Galmaeki commented on src/main/java/com/final_10aeat/domain/member/service/MemberService.java:
```diff
@@ -20,30 +26,38 @@ public class MemberService {
     @Transactional(readOnly = true)
     public String login(MemberLoginRequestDto request) {
-        Member member = memberRepository.findByEmail(request.email())
+        Member member = memberRepository.findByEmailAndDeletedAtIsNull(request.email())
```
서버 최신상황 반영시엔 테스트코드의 given 부분이 변경될겁니다!

Galmaeki reviewed (changes requested):
고생하셨슴다!
댓글로 남겨놨으니 확인 부탁드림다!

asd42270 committed:
Merge remote-tracking branch 'origin/feature/63' into feature/63

# Conflicts:
#	src/main/java/com/final_10aeat/domain/member/controller/MemberController.java

asd42270 committed:
[Refactor] PR 수정 사항 반영
- MemberWithdrawRequestDto: Validation 추가
 - MemberServiceTest: deleteAt 필터링 추가로 인한 메서드 변경 사항 반영
 - MemberController: 기존 사항 재커밋
 - WithdrawServiceTest: 테스트 코드 수정 중...내일 아침에 다시 수정하겠습니다.

Galmaeki commented on src/main/java/com/final_10aeat/domain/member/controller/MemberController.java:
```diff
@@ -29,9 +27,17 @@ public ResponseDTO<Void> register(
     public ResponseDTO<Void> login(
             HttpServletResponse response,
-            @RequestBody @Valid MemberLoginRequestDto request) {
+            @RequestBody MemberLoginRequestDto request) {
```
@Valid..... 😢

asd42270 committed:
[Refactor] PR 수정 사항 반영
- MemberController: @Valid 어노테이션 추가
 - MemberRegisterRequestDto: Validation 추가
 - MemberWithdrawRequestDto: Validation 추가
 - WithdrawServiceTest: 테스트코드 setUp()에서 when -> given으로 변경

asd42270 committed:
[Refactor] PR 수정 사항 반영
- MemberController: 회원가입 완료시 로그인으로 넘어가도록 코드 수정
 - MemberService: 회원가입 완료 시 로그인 실행을 위해 return 타입 MemberLoginRequestDto 추가
 - RegisterServiceTest: register 테스트 코드 추가
 - WithdrawServiceTest: given(existsByEmailAndDeletedAtIsNull)의 willReturn을 false로 지정

asd42270 committed:
[Refactor] PR 수정 사항 반영
- MemberController: 사용하지 않는 괄호 삭제, @Valid 추가

Galmaeki reviewed (approved):
고생 많으셨습니다!

ghrltjdtprbs reviewed (approved):
변경사항 확인했습니다! 수고하셨어요 ㅎㅎ 👍👍
SecurityConfig에 회원가입 경로 변경사항만 반영해주시면 될 것 같습니다!

asd42270 committed:
[Refactor] PR 수정 사항 반영
- 충돌 사항 반영
- 약관 동의 여부 컬럼추가
- 그에 따른 서비스 로직 변경

asd42270 committed:
Merge branch 'develop' into feature/63
